contracts/AelinPool.sol

```diff
@@ -385,6 +445,83 @@ contract AelinPool is AelinERC20, MinimalProxyFactory, IAelinPool {
         emit PurchasePoolToken(msg.sender, purchaseTokenAmount);
     }
 
+    /**
+     * @dev allows anyone to become a purchaser with a qualified erc721
+     * nft in the pool depending on the scenarios
+     *
+     * Scenarios:
+     * 1. each wallet holding a qualified NFT to deposit an unlimited amount of purchase tokens
+     * 2. certain amount of purchase tokens per wallet regardless of the number of qualified NFTs held
+     * 3. certain amount of Investment tokens per qualified NFT held
+     */
+
+    function purchasePoolTokensWithNft(
+        address _collectionAddress,
+        uint256[] calldata _tokenIds,
+        uint256 _purchaseTokenAmount
+    ) external lock {
+        require(hasNftList, "pool does not have an NFT list");
+        require(block.timestamp < purchaseExpiry, "not in purchase window");
+        require(!nftWalletUsedForPurchase[msg.sender], "wallet already used");
+
+        NftData memory tmpNftData = nftCollectionDetails[_collectionAddress];
+        require(tmpNftData.collectionAddress == _collectionAddress, "collection not in the pool");
+
+        if (tmpNftData.purchaseAmountPerToken) {
+            require(
+                _purchaseTokenAmount <= tmpNftData.purchaseAmount * _tokenIds.length,
+                "purchase amount per token should be less than the allocation"
+            );
+        }
+
+        if (!tmpNftData.purchaseAmountPerToken && tmpNftData.purchaseAmount > 0) {
+            require(
+                nftAllowList[msg.sender] + _purchaseTokenAmount <= tmpNftData.purchaseAmount,
+                "purchase amount should be less than the allocation"
+            );
+            nftAllowList[msg.sender] += _purchaseTokenAmount;
+        }
+
+        if (NftCheck.supports721(_collectionAddress)) _purchasePoolWith721(_collectionAddress, _tokenIds);
+        if (NftCheck.supports1155(_collectionAddress)) _purchasePoolWith1155(_collectionAddress, _tokenIds, tmpNftData);
+        if (_collectionAddress == CRYPTO_PUNKS) _purchasePoolWithPunks(_collectionAddress, _tokenIds);
+
+        nftWalletUsedForPurchase[msg.sender] = true;
+        IERC20(purchaseToken).safeTransferFrom(msg.sender, address(this), _purchaseTokenAmount);
```

Comment:
another thing we have to consider for token transfers is if a token burns a portion of its supply during a transfer. this is not how we want to do the accounting

Reply:
uint256 amountBefore = IERC20(purchaseToken).balanceOf(address(this));
IERC20(purchaseToken).safeTransferFrom(msg.sender, address(this), _purchaseTokenAmount);
uint256 amountAfter = IERC20(purchaseToken).balanceOf(address(this));
uint256 amountSent = amountAfter - amountBefore;

NOTE I am just typing this without testing but we want something like this instead to protect ourselves

Reply:
yea i wanted to ask you about this and wanted to know why you used an approach like this. can you please explain what do you mean by a portion of the token is burnt during transfer?

> uint256 amountBefore = IERC20(purchaseToken).balanceOf(address(this)); IERC20(purchaseToken).safeTransferFrom(msg.sender, address(this), _purchaseTokenAmount); uint256 amountAfter = IERC20(purchaseToken).balanceOf(address(this)); uint256 amountSent = amountAfter - amountBefore;
> 
> NOTE I am just typing this without testing but we want something like this instead to protect ourselves

Reply:
done. 5e0426fb0e68557fda5f49de0412911a8bbed85b